packages/smithy-core/src/smithy_core/documents.py

```diff
@@ -633,3 +638,24 @@ def read_document(self, schema: "Schema") -> Document:
     @override
     def read_timestamp(self, schema: "Schema") -> datetime.datetime:
         return self._value.as_timestamp()
+
+
+# A registry for on-demand deserialization of types by using a mapping of shape IDs to their deserializers.
+class TypeRegistry:
+    def __init__(
+        self,
+        types: dict[ShapeID, type[DeserializeableShape]],
+        sub_registry: "TypeRegistry | None" = None,
+    ):
+        self._types = types
+        self._sub_registry = sub_registry
+
+    def get(self, shape: ShapeID) -> type[DeserializeableShape]:
```

Comment:
This class should probably also implement `__getitem__` and `__contains__`. Also these methods need docs.